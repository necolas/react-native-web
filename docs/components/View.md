# View

`View` is the fundamental UI building block. It is a component that supports
style, layout with flexbox, and accessibility controls. It can be nested
inside another `View` and has 0-to-many children of any type.

Also, refer to React Native's documentation about the [Gesture Responder
System](http://facebook.github.io/react-native/releases/0.22/docs/gesture-responder-system.html).

Unsupported React Native props:
`onAccessibilityTap`,
`hitSlop`,
`onMagicTap`

## Props

NOTE: `View` will transfer all other props to the rendered HTML element.

**accessibilityLabel**: string

Defines the text available to assistive technologies upon interaction with the
element. (This is implemented using `aria-label`.)

**accessibilityLiveRegion**: oneOf('assertive', 'off', 'polite') = 'off'

Indicates to assistive technologies whether to notify the user when the view
changes. The values of this attribute are expressed in degrees of importance.
When regions are specified as `polite` (recommended), updates take low
priority. When regions are specified as `assertive`, assistive technologies
will interrupt and immediately notify the user. (This is implemented using
[`aria-live`](http://www.w3.org/TR/wai-aria/states_and_properties#aria-live).)

(web) **accessibilityRole**: oneOf(roles)

Allows assistive technologies to present and support interaction with the view
in a manner that is consistent with user expectations for similar views of that
type. For example, marking a touchable view with an `accessibilityRole` of
`button`. (This is implemented using [ARIA roles](http://www.w3.org/TR/wai-aria/roles#role_definitions)).

Note: Avoid changing `accessibilityRole` values over time or after user
actions. Generally, accessibility APIs do not provide a means of notifying
assistive technologies of a `role` value change.

**accessible**: bool = true

When `false`, the view is hidden from assistive technologies. (This is
implemented using `aria-hidden`.)

**onLayout**: function

Invoked on mount and layout changes with `{ nativeEvent: { layout: { x, y, width,
height } } }`, where `x` and `y` are the offsets from the parent node.

**onMoveShouldSetResponder**: function

**onMoveShouldSetResponderCapture**: function

**onResponderGrant**: function

For most touch interactions, you'll simply want to wrap your component in
`TouchableHighlight` or `TouchableOpacity`.

**onResponderMove**: function

**onResponderReject**: function

**onResponderRelease**: function

**onResponderTerminate**: function

**onResponderTerminationRequest**: function

**onStartShouldSetResponder**: function

**onStartShouldSetResponderCapture**: function

**pointerEvents**: oneOf('auto', 'box-only', 'box-none', 'none') = 'auto'

Configure the `pointerEvents` of the view. The enhanced `pointerEvents` modes
provided are not part of the CSS spec, therefore, `pointerEvents` is excluded
from `style`.

`box-none` is the equivalent of:

```css
.box-none { pointer-events: none }
.box-none * { pointer-events: auto }
```

`box-only` is the equivalent of:

```css
.box-only { pointer-events: auto }
.box-only * { pointer-events: none }
```

**style**: style

+ `alignContent`
+ `alignItems`
+ `alignSelf`
+ `animationDelay`
+ `animationDirection`
+ `animationDuration`
+ `animationFillMode`
+ `animationIterationCount`
+ `animationName`
+ `animationPlayState`
+ `animationTimingFunction`
+ `backfaceVisibility`
+ `backgroundAttachment`
+ `backgroundClip`
+ `backgroundColor`
+ `backgroundImage`
+ `backgroundOrigin`
+ `backgroundPosition`
+ `backgroundRepeat`
+ `backgroundSize`
+ `borderColor` (single value)
+ `borderTopColor`
+ `borderBottomColor`
+ `borderRightColor`‡
+ `borderLeftColor`‡
+ `borderRadius` (single value)
+ `borderTopLeftRadius`‡
+ `borderTopRightRadius`‡
+ `borderBottomLeftRadius`‡
+ `borderBottomRightRadius`‡
+ `borderStyle` (single value)
+ `borderTopStyle`
+ `borderRightStyle`‡
+ `borderBottomStyle`
+ `borderLeftStyle`‡
+ `borderWidth` (single value)
+ `borderBottomWidth`
+ `borderLeftWidth`‡
+ `borderRightWidth`‡
+ `borderTopWidth`
+ `bottom`
+ `boxShadow`
+ `boxSizing`
+ `cursor`
+ `flex` (number)
+ `flexBasis`
+ `flexDirection`
+ `flexGrow`
+ `flexShrink`
+ `flexWrap`
+ `height`
+ `justifyContent`
+ `left`‡
+ `margin` (single value)
+ `marginBottom`
+ `marginHorizontal`
+ `marginLeft`‡
+ `marginRight`‡
+ `marginTop`
+ `marginVertical`
+ `maxHeight`
+ `maxWidth`
+ `minHeight`
+ `minWidth`
+ `opacity`
+ `order`
+ `overflow`
+ `overflowX`
+ `overflowY`
+ `padding` (single value)
+ `paddingBottom`
+ `paddingHorizontal`
+ `paddingLeft`‡
+ `paddingRight`‡
+ `paddingTop`
+ `paddingVertical`
+ `perspective`
+ `perspectiveOrigin`
+ `position`
+ `right`‡
+ `top`
+ `transform`
+ `transformOrigin`
+ `transitionDelay`
+ `transitionDuration`
+ `transitionProperty`
+ `transitionTimingFunction`
+ `userSelect`
+ `visibility`
+ `width`
+ `zIndex`

‡ This property can be suffixed with `$noI18n` to prevent automatic
bidi-flipping in RTL mode. This is only supported if `Platform.OS === 'web'`.

Default:

```js
{
  alignItems: 'stretch',
  borderWidth: 0,
  borderStyle: 'solid',
  boxSizing: 'border-box',
  display: 'flex',
  flexBasis: 'auto',
  flexDirection: 'column',
  flexShrink: 0,
  margin: 0,
  padding: 0,
  position: 'relative'
};
```

(See [facebook/css-layout](https://github.com/facebook/css-layout)).

**testID**: string

Used to locate this view in end-to-end tests.

## Examples

```js
import React, { Component, PropTypes } from 'react'
import { StyleSheet, View } from 'react-native'

export default class ViewExample extends Component {
  render() {
    return (
      <View style={styles.row}>
        {
          ['1', '2', '3', '4', '5'].map((value, i) => {
            return <View children={value} key={i} style={styles.cell} />
          })
        }
      </View>
    );
  }
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row'
  },
  cell: {
    flexGrow: 1
  }
})
```
