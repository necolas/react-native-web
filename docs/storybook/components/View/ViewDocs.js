/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import PropPointerEvents from './examples/PropPointerEvents';
import transformExamples from './examples/transforms';
import ZIndexExample from './examples/ZIndex';
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import UIExplorer, { AppText, Code, DocItem, StyleList } from '../../ui-explorer';

const stylePropTypes = [
  {
    name: 'alignContent',
    typeInfo: 'string'
  },
  {
    name: 'alignItems',
    typeInfo: 'string'
  },
  {
    name: 'alignSelf',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationDelay',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationDirection',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationDuration',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationFillMode',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationName',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationIterationCount',
    typeInfo: 'number | "infinite"'
  },
  {
    label: 'web',
    name: 'animationPlayState',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'animationTimingFunction',
    typeInfo: 'string'
  },
  {
    name: 'backfaceVisibility',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundAttachment',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundBlendMode',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundClip',
    typeInfo: 'string'
  },
  {
    name: 'backgroundColor',
    typeInfo: 'color'
  },
  {
    label: 'web',
    name: 'backgroundImage',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundOrigin',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundPosition',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundRepeat',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'backgroundSize',
    typeInfo: 'string'
  },
  {
    name: 'borderColor',
    typeInfo: 'color'
  },
  {
    name: 'borderTopColor',
    typeInfo: 'color'
  },
  {
    name: 'borderBottomColor',
    typeInfo: 'color'
  },
  {
    name: 'borderRightColor',
    typeInfo: 'color'
  },
  {
    name: 'borderLeftColor',
    typeInfo: 'color'
  },
  {
    name: 'borderRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'borderTopLeftRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'borderTopRightRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'borderBottomLeftRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'borderBottomRightRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'borderStyle',
    typeInfo: 'string'
  },
  {
    name: 'borderTopStyle',
    typeInfo: 'string'
  },
  {
    name: 'borderRightStyle',
    typeInfo: 'string'
  },
  {
    name: 'borderBottomStyle',
    typeInfo: 'string'
  },
  {
    name: 'borderLeftStyle',
    typeInfo: 'string'
  },
  {
    name: 'borderWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'borderBottomWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'borderLeftWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'borderRightWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'borderTopWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'bottom',
    typeInfo: 'number | string'
  },
  {
    label: 'web',
    name: 'boxShadow',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'boxSizing',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'clip',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'cursor',
    typeInfo: 'string'
  },
  {
    name: 'direction',
    typeInfo: 'string'
  },
  {
    name: 'display',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'filter',
    typeInfo: 'string'
  },
  {
    name: 'flex',
    typeInfo: 'number'
  },
  {
    name: 'flexBasis',
    typeInfo: 'string'
  },
  {
    name: 'flexDirection',
    typeInfo: 'string'
  },
  {
    name: 'flexGrow',
    typeInfo: 'number'
  },
  {
    name: 'flexShrink',
    typeInfo: 'number'
  },
  {
    name: 'flexWrap',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridAutoColumns',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridAutoFlow',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridAutoRows',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridColumnEnd',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridColumnGap',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridColumnStart',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridRowEnd',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridRowGap',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridRowStart',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridTemplateColumns',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridTemplateRows',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'gridTemplateAreas',
    typeInfo: 'string'
  },
  {
    name: 'height',
    typeInfo: 'number | string'
  },
  {
    name: 'justifyContent',
    typeInfo: 'string'
  },
  {
    name: 'left',
    typeInfo: 'number | string'
  },
  {
    name: 'margin',
    typeInfo: 'number | string'
  },
  {
    name: 'marginBottom',
    typeInfo: 'number | string'
  },
  {
    name: 'marginHorizontal',
    typeInfo: 'number | string'
  },
  {
    name: 'marginLeft',
    typeInfo: 'number | string'
  },
  {
    name: 'marginRight',
    typeInfo: 'number | string'
  },
  {
    name: 'marginTop',
    typeInfo: 'number | string'
  },
  {
    name: 'marginVertical',
    typeInfo: 'number | string'
  },
  {
    name: 'maxHeight',
    typeInfo: 'number | string'
  },
  {
    name: 'maxWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'minHeight',
    typeInfo: 'number | string'
  },
  {
    name: 'minWidth',
    typeInfo: 'number | string'
  },
  {
    name: 'opacity',
    typeInfo: 'number'
  },
  {
    name: 'order',
    typeInfo: 'number'
  },
  {
    label: 'web',
    name: 'outline',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'outlineColor',
    typeInfo: 'color'
  },
  {
    name: 'overflow',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'overflowX',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'overflowY',
    typeInfo: 'string'
  },
  {
    name: 'padding',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingBottom',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingHorizontal',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingLeft',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingRight',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingTop',
    typeInfo: 'number | string'
  },
  {
    name: 'paddingVertical',
    typeInfo: 'number | string'
  },
  {
    label: 'web',
    name: 'perspective',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'perspectiveOrigin',
    typeInfo: 'string'
  },
  {
    name: 'position',
    typeInfo: 'string'
  },
  {
    name: 'right',
    typeInfo: 'number | string'
  },
  {
    name: 'shadowColor',
    typeInfo: 'color'
  },
  {
    name: 'shadowOffset',
    typeInfo: '{ height: number | string, width: number | string }'
  },
  {
    name: 'shadowOpacity',
    typeInfo: 'number'
  },
  {
    name: 'shadowRadius',
    typeInfo: 'number | string'
  },
  {
    name: 'top',
    typeInfo: 'number | string'
  },
  {
    name: 'transform',
    typeInfo: 'array'
  },
  {
    label: 'web',
    name: 'transformOrigin',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'transitionDelay',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'transitionDuration',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'transitionProperty',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'transitionTimingFunction',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'userSelect',
    typeInfo: 'string'
  },
  {
    label: 'web',
    name: 'visibility',
    typeInfo: 'string'
  },
  {
    name: 'width',
    typeInfo: 'number | string'
  },
  {
    label: 'web',
    name: 'willChange',
    typeInfo: 'string'
  },
  {
    name: 'zIndex',
    typeInfo: 'number'
  }
];

const sections = [
  {
    title: 'Props',
    entries: [
      <DocItem
        name="accessibilityLabel"
        typeInfo="?string"
        description="Overrides the text that's read by a screen reader when the user interacts with the element. (This is implemented using 'aria-label'.)"
      />,

      <DocItem
        name="accessibilityLiveRegion"
        typeInfo="?enum('assertive', 'none', 'polite')"
        description={
          <AppText>
            Indicates to assistive technologies whether to notify the user when the view changes.
            The values of
            this attribute are expressed in degrees of importance. When regions are specified as{
              ' '
            }
            <Code>polite</Code> (recommended),
            updates take low priority. When regions are specified as <Code>assertive</Code>,
            assistive technologies will
            interrupt and immediately notify the user. (This is implemented using 'aria-live'.)
          </AppText>
        }
      />,

      <DocItem
        label="web"
        name="accessibilityRole"
        typeInfo="?enum(roles)"
        description={
          <AppText>
            Allows assistive technologies to present and support interaction with the view in a
            manner that is
            consistent with user expectations for similar views of that type. For example, marking a
            touchable
            view with an <Code>accessibilityRole</Code> of <Code>button</Code>. For compatibility
            with React
            Native <Code>accessibilityTraits</Code> and <Code>accessibilityComponentType</Code> are
            mapped
            to <Code>accessibilityRole</Code>. (This is implemented using ARIA roles.)
          </AppText>
        }
      />,

      <DocItem
        name="accessible"
        typeInfo="?boolean"
        description={
          <AppText>
            When <Code>true</Code>, indicates that the view is an accessibility element (i.e.,
            focusable) and
            groups its child content. By default, all the touchable elements and elements
            with <Code>accessibilityRole</Code> of <Code>button</Code> and <Code>link</Code> are
            accessible.
            (This is implemented using 'tabindex'.)
          </AppText>
        }
      />,

      <DocItem name="children" typeInfo="?element" description="Child content" />,

      <DocItem
        name="hitSlop"
        typeInfo="?object"
        description={
          'This defines how far a touch event can start away from the view (in pixels). Typical interface guidelines recommend touch targets that are at least 30 - 40 points/density-independent pixels.\n\nFor example, if a touchable view has a height of `20` the touchable height can be extended to `40` with hitSlop.'
        }
        example={{
          code: '<View hitSlop={{top: 10, bottom: 10, left: 0, right: 0}} />',
          render: () => null
        }}
      />,

      <DocItem
        name="importantForAccessibility"
        typeInfo="?enum('auto', 'no', 'no-hide-descendants', 'yes')"
        description={
          'A value of `no` will remove the element from the tab flow.\n\nA value of `no-hide-descendants` will hide the element and its children from assistive technologies. (This is implemented using `aria-hidden`.)'
        }
      />,

      <DocItem
        name="onLayout"
        typeInfo="?function"
        description="Invoked on mount and layout changes with `{ nativeEvent: { layout: { x, y, width, height } } }`, where `x` and `y` are the offsets from the parent node."
      />,

      <DocItem
        name="onMoveShouldSetResponder"
        typeInfo="?function => boolean"
        description="Does this view want to &quot;claim&quot; touch responsiveness? This is called for every touch move on the `View` when it is not the responder."
      />,

      <DocItem
        name="onMoveShouldSetResponderCapture"
        typeInfo="?function => boolean"
        description="If a parent `View` wants to prevent a child `View` from becoming responder on a move, it should have this handler return `true`."
      />,

      <DocItem
        name="onResponderGrant"
        typeInfo="?function"
        description="The `View` is now responding to touch events. This is the time to highlight and show the user what is happening. For most touch interactions, you'll simply want to wrap your component in `TouchableHighlight` or `TouchableOpacity`."
      />,

      <DocItem
        name="onResponderMove"
        typeInfo="?function"
        description="The user is moving their finger."
      />,

      <DocItem
        name="onResponderReject"
        typeInfo="?function"
        description="Another responder is already active and will not release it to the `View` asking to be the responder."
      />,

      <DocItem
        name="onResponderRelease"
        typeInfo="?function"
        description="Fired at the end of the touch."
      />,

      <DocItem
        name="onResponderTerminate"
        typeInfo="?function"
        description="The responder has been taken from the `View`."
      />,

      <DocItem
        name="onResponderTerminationRequest"
        typeInfo="?function"
        description="Some other `View` wants to become responder and is asking this `View` to release its responder. Returning `true` allows its release."
      />,

      <DocItem
        name="onStartShouldSetResponder"
        typeInfo="?function => boolean"
        description="Does this view want to become responder on the start of a touch?"
      />,

      <DocItem
        name="onStartShouldSetResponderCapture"
        typeInfo="?function => boolean"
        description="If a parent `View` wants to prevent a child `View` from becoming the responder on a touch start, it should have this handler return `true`."
      />,

      <DocItem
        name="pointerEvents"
        typeInfo="?enum('auto', 'box-only', 'box-none', 'none') = 'auto'"
        description="Controls whether the View can be the target of touch events. The enhanced `pointerEvents` modes provided are not part of the CSS spec, therefore, `pointerEvents` is excluded from `style`. `box-none` preserves pointer events on the element's children; `box-only` disables pointer events on the element's children"
        example={{
          code: '',
          render: () => <PropPointerEvents />
        }}
      />,

      <DocItem
        name="style"
        typeInfo="?style"
        description={<StyleList stylePropTypes={stylePropTypes} />}
      />,

      <DocItem
        name="testID"
        typeInfo="?string"
        description="Used to locate this view in end-to-end tests. The test ID is rendered to a `data-testid` DOM attribute"
      />,

      <DocItem
        label="compat"
        name="accessibilityComponentType"
        typeInfo="?enum(roles)"
        description="(For compatibility with React Native. Equivalent to &quot;accessibilityRole&quot;.)"
      />,

      <DocItem
        label="compat"
        name="accessibilityTraits"
        typeInfo="?enum(roles) | Array<role>"
        description="(For compatibility with React Native. Equivalent to &quot;accessibilityRole&quot;.)"
      />
    ]
  },

  {
    title: 'More examples',
    entries: [
      <DocItem
        description="z-index"
        example={{
          render: () => <ZIndexExample />
        }}
      />
    ].concat(
      transformExamples.map(({ title, render }) =>
        <DocItem description={title} example={{ render }} />
      )
    )
  }
];

storiesOf('Components', module).add('View', () =>
  <UIExplorer
    description={[
      <AppText>
        View is the fundamental UI building block. It is a component that supports style, layout
        with flexbox,
        and accessibility controls. It can be nested inside another View and has 0-to-many children
        of any type.
      </AppText>,
      <AppText>
        Also, refer to React Native's documentation about the Gesture Responder System. NOTE: View
        will
        transfer all other props to the rendered HTML element.
      </AppText>
    ]}
    sections={sections}
    title="View"
    url="components/View"
  />
);
