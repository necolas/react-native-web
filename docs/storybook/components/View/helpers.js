/**
 * @flow
 */

export const logger = e => {
  console.log(e.nativeEvent);
};
