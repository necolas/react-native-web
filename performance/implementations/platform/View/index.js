/* eslint-disable react/prop-types */
import classnames from 'classnames';
import React from 'react';

const View = props => <div {...props} className={classnames('view', props.className)} />;

module.exports = View;
