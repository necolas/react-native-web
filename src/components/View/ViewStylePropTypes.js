import AnimationPropTypes from '../../propTypes/AnimationPropTypes';
import BorderPropTypes from '../../propTypes/BorderPropTypes';
import ColorPropType from '../../propTypes/ColorPropType';
import LayoutPropTypes from '../../propTypes/LayoutPropTypes';
import { PropTypes } from 'react';
import ShadowPropTypes from '../../propTypes/ShadowPropTypes';
import TransformPropTypes from '../../propTypes/TransformPropTypes';

const { number, oneOf, string } = PropTypes;
const autoOrHiddenOrVisible = oneOf([ 'auto', 'hidden', 'visible' ]);
const hiddenOrVisible = oneOf([ 'hidden', 'visible' ]);

module.exports = process.env.NODE_ENV !== 'production' ? {
  ...AnimationPropTypes,
  ...BorderPropTypes,
  ...LayoutPropTypes,
  ...ShadowPropTypes,
  ...TransformPropTypes,
  backfaceVisibility: hiddenOrVisible,
  backgroundColor: ColorPropType,
  opacity: number,
  overflow: autoOrHiddenOrVisible,
  zIndex: number,
  /*
   * @platform web
   */
  backgroundAttachment: string,
  backgroundClip: string,
  backgroundImage: string,
  backgroundOrigin: oneOf([ 'border-box', 'content-box', 'padding-box' ]),
  backgroundPosition: string,
  backgroundRepeat: string,
  backgroundSize: string,
  boxShadow: string,
  cursor: string,
  display: string,
  outline: string,
  overflowX: autoOrHiddenOrVisible,
  overflowY: autoOrHiddenOrVisible,
  perspective: PropTypes.oneOfType([ number, string ]),
  perspectiveOrigin: string,
  transitionDelay: string,
  transitionDuration: string,
  transitionProperty: string,
  transitionTimingFunction: string,
  userSelect: string,
  visibility: hiddenOrVisible,
  WebkitOverflowScrolling: oneOf([ 'auto', 'touch' ])
} : {};
