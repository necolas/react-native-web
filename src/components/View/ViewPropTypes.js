/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule ViewPropTypes
 * @flow
 */

import BaseComponentPropTypes from '../../propTypes/BaseComponentPropTypes';
import EdgeInsetsPropType from '../../propTypes/EdgeInsetsPropType';
import StyleSheetPropType from '../../propTypes/StyleSheetPropType';
import ViewStylePropTypes from './ViewStylePropTypes';
import { any, bool, func, oneOf } from 'prop-types';

const ViewPropTypes = {
  ...BaseComponentPropTypes,
  children: any,
  collapsable: bool,
  hitSlop: EdgeInsetsPropType,
  onClick: func,
  onClickCapture: func,
  onLayout: func,
  onMoveShouldSetResponder: func,
  onMoveShouldSetResponderCapture: func,
  onResponderGrant: func,
  onResponderMove: func,
  onResponderReject: func,
  onResponderRelease: func,
  onResponderTerminate: func,
  onResponderTerminationRequest: func,
  onStartShouldSetResponder: func,
  onStartShouldSetResponderCapture: func,
  onTouchCancel: func,
  onTouchCancelCapture: func,
  onTouchEnd: func,
  onTouchEndCapture: func,
  onTouchMove: func,
  onTouchMoveCapture: func,
  onTouchStart: func,
  onTouchStartCapture: func,
  pointerEvents: oneOf(['auto', 'box-none', 'box-only', 'none']),
  style: StyleSheetPropType(ViewStylePropTypes)
};

export default ViewPropTypes;
