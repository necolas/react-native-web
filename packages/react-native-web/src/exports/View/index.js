/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * @providesModule View
 * @flow
 */

import applyLayout from '../../modules/applyLayout';
import applyNativeMethods from '../../modules/applyNativeMethods';
import { bool } from 'prop-types';
import createElement from '../createElement';
import invariant from 'fbjs/lib/invariant';
import StyleSheet from '../StyleSheet';
import ViewPropTypes, { type ViewProps } from './ViewPropTypes';
import React, { Component } from 'react';

const calculateHitSlopStyle = hitSlop => {
  const hitStyle = {};
  for (const prop in hitSlop) {
    if (hitSlop.hasOwnProperty(prop)) {
      const value = hitSlop[prop];
      hitStyle[prop] = value > 0 ? -1 * value : 0;
    }
  }
  return hitStyle;
};

class View extends Component<ViewProps> {
  static displayName = 'View';

  static contextTypes = {
    isInAParentText: bool
  };

  static propTypes = ViewPropTypes;

  render() {
    const {
      hitSlop,
      /* eslint-disable */
      accessibilityViewIsModal,
      collapsable,
      needsOffscreenAlphaCompositing,
      onAccessibilityTap,
      onLayout,
      onMagicTap,
      removeClippedSubviews,
      renderToHardwareTextureAndroid,
      shouldRasterizeIOS,
      /* eslint-enable */
      ...otherProps
    } = this.props;

    if (process.env.NODE_ENV !== 'production') {
      React.Children.toArray(this.props.children).forEach(item => {
        invariant(
          typeof item !== 'string',
          `Unexpected text node: ${item}. A text node cannot be a child of a <View>.`
        );
      });
    }

    const { isInAParentText } = this.context;

    otherProps.style = StyleSheet.compose(
      styles.initial,
      StyleSheet.compose(isInAParentText && styles.inline, this.props.style)
    );

    if (hitSlop) {
      const hitSlopStyle = calculateHitSlopStyle(hitSlop);
      const hitSlopChild = createElement('span', { style: [styles.hitSlop, hitSlopStyle] });
      otherProps.children = React.Children.toArray([hitSlopChild, otherProps.children]);
    }

    return createElement('div', otherProps);
  }
}

const styles = StyleSheet.create({
  // https://github.com/facebook/css-layout#default-values
  initial: {
    alignItems: 'stretch',
    borderWidth: 0,
    borderStyle: 'solid',
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    margin: 0,
    padding: 0,
    position: 'relative',
    zIndex: 0,
    // fix flexbox bugs
    minHeight: 0,
    minWidth: 0
  },
  inline: {
    display: 'inline-flex'
  },
  // this zIndex-ordering positions the hitSlop above the View but behind
  // its children
  hitSlop: {
    ...StyleSheet.absoluteFillObject,
    zIndex: -1
  }
});

export default applyLayout(applyNativeMethods(View));
