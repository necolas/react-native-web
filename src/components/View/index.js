/**
 * Copyright (c) 2015-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * @providesModule View
 * @flow
 */

import applyLayout from '../../modules/applyLayout';
import applyNativeMethods from '../../modules/applyNativeMethods';
import { bool } from 'prop-types';
import createElement from '../../modules/createElement';
import invariant from 'fbjs/lib/invariant';
import StyleSheet from '../../apis/StyleSheet';
import ViewPropTypes from './ViewPropTypes';
import React, { Component } from 'react';

const calculateHitSlopStyle = hitSlop => {
  const hitStyle = {};
  for (const prop in hitSlop) {
    if (hitSlop.hasOwnProperty(prop)) {
      const value = hitSlop[prop];
      hitStyle[prop] = value > 0 ? -1 * value : 0;
    }
  }
  return hitStyle;
};

class View extends Component {
  static displayName = 'View';

  static contextTypes = {
    isInAParentText: bool
  };

  static propTypes = ViewPropTypes;

  render() {
    const {
      hitSlop,
      style,
      /* eslint-disable */
      collapsable,
      onAccessibilityTap,
      onLayout,
      onMagicTap,
      removeClippedSubviews,
      /* eslint-enable */
      ...otherProps
    } = this.props;

    if (process.env.NODE_ENV !== 'production') {
      React.Children.toArray(this.props.children).forEach(item => {
        invariant(
          typeof item !== 'string',
          `Unexpected text node: ${item}. A text node cannot be a child of a <View>.`
        );
      });
    }

    const { isInAParentText } = this.context;

    otherProps.style = [styles.initial, isInAParentText && styles.inline, style];

    if (hitSlop) {
      const hitSlopStyle = calculateHitSlopStyle(hitSlop);
      const hitSlopChild = createElement('span', { style: [styles.hitSlop, hitSlopStyle] });
      otherProps.children = React.Children.toArray(otherProps.children);
      otherProps.children.unshift(hitSlopChild);
      otherProps.style.unshift(styles.hasHitSlop);
    }

    return createElement('div', otherProps);
  }
}

const styles = StyleSheet.create({
  // https://github.com/facebook/css-layout#default-values
  initial: {
    alignItems: 'stretch',
    borderWidth: 0,
    borderStyle: 'solid',
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    margin: 0,
    padding: 0,
    position: 'relative',
    // fix flexbox bugs
    minHeight: 0,
    minWidth: 0
  },
  inline: {
    display: 'inline-flex'
  },
  // this zIndex-ordering positions the hitSlop above the View but behind
  // its children
  hasHitSlop: {
    zIndex: 0
  },
  hitSlop: {
    ...StyleSheet.absoluteFillObject,
    zIndex: -1
  }
});

export default applyLayout(applyNativeMethods(View));
