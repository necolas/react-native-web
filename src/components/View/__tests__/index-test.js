/* eslint-env jasmine, jest */

import React from 'react';
import { render, shallow } from 'enzyme';
import View from '../';

describe('components/View', () => {
  describe('rendered element', () => {
    test('is a "div" by default', () => {
      const component = shallow(<View />);
      expect(component.type()).toBe('div');
    });
  });

  test('prop "children"', () => {
    const children = <View testID="1" />;
    const component = shallow(<View>{children}</View>);
    expect(component.contains(children)).toEqual(true);
  });

  describe('prop "hitSlop"', () => {
    it('renders a span with negative position offsets', () => {
      const component = render(<View hitSlop={{ top: 10, bottom: 10, right: 5, left: 5 }} />);
      expect(component).toMatchSnapshot();
    });

    it('handles partial offsets', () => {
      const component = render(<View hitSlop={{ top: 10 }} />);
      expect(component).toMatchSnapshot();
    });
  });

  test('prop "pointerEvents"', () => {
    const component = render(<View pointerEvents="box-only" />);
    expect(component).toMatchSnapshot();
  });
});
