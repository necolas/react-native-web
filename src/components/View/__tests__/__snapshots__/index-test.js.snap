// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/View prop "hitSlop" handles partial offsets 1`] = `
Object {
  "top": "-10px",
}
`;

exports[`components/View prop "hitSlop" renders a span with negative position offsets 1`] = `
Object {
  "bottom": "-10px",
  "left": "-5px",
  "right": "-5px",
  "top": "-10px",
}
`;
